### PR TITLE
MAILET-115 Refactor AbstractRedirect & AbstractNotify based mailets

### PR DESCRIPTION
**Pull request by aduprat**
#423 + 6 last commits.

### CODE REVIEW

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/Resend.java [hunk 1]
```diff
@@ -323,4 +324,20 @@ protected String getMessage(Mail originalMail) throws MessagingException {
...
+    protected MailAddress getReplyTo() throws MessagingException {
+        String replyTo = getInitParameters().getReplyTo();
+        if (Strings.isNullOrEmpty(replyTo)) {
+            return null;
+        }
+
+        List<MailAddress> extractAddresses = AddressExtractor.withContext(getMailetContext())
+                .allowedSpecials(ImmutableList.of("postmaster", "sender", "null", "unaltered"))
+                .extract(replyTo);
+        if (extractAddresses.isEmpty()) {
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/Resend.java]: This pattern of returning first element seems recurent. Maybe you can add a **extractSingleAddress** method to AddressExtractor

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AbstractRedirect.java [hunk 2]
```diff
@@ -475,11 +474,33 @@ protected void setSender(Mail newMail, MailAddress sender, Mail originalMail) th
      * <i>originalMail</i> to <i>subjectPrefix</i>. Is a "setX(Mail, Tx, Mail)"
      * method.
      */
-    protected void setSubjectPrefix(Mail newMail, String subjectPrefix, Mail originalMail) throws MessagingException {
-        if (isNotifyMailet()) {
-            new MimeMessageModifier(originalMail.getMessage()).addSubjectPrefix(subjectPrefix);
-        }
-
+    protected abstract void setSubjectPrefix(Mail newMail, String subjectPrefix, Mail originalMail) throws MessagingException;
+//        if (isNotifyMailet()) {
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/redirect/AbstractRedirect.java]: Commented code is not needed with SCM tools

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Pull request closed without merging**